Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift

```diff
@@ -761,13 +762,13 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
 
     internal func populateTargetTrie(callback: @escaping (Result<Void, Error>) -> Void = { _ in }) {
         DispatchQueue.sharedConcurrent.async(group: nil, qos: .background, flags: .assignCurrentContext) {
-            self.targetTrieReady.memoize {
-                do {
+            do {
+                try self.populateTargetTrieLock.withLock { // Prevent race to populate targetTrie
```

Comment:
this can exit early if  self.targetTrieReady = true

Reply:
`self.targetTrieReady = true` only happens at the end of the block though. We exit early on L770.

Reply:
agreed. what I meant is that if populateTargetTrie is called from two threads concurrently, the lock will prevent it from running in parallel (good) but won't prevent it from running twice (probably safe but wasteful), so if we checked on if populateTargetTrie is true already right after the lock we can save that second computation.

Reply:
i see. added https://github.com/apple/swift-package-manager/pull/3522/commits/b47deb72dc8bacb38a8ac89d93fc20cd22612c66 so we only compute once.